src/lib/pubkey/hss_lms/hss.cpp

```diff
@@ -0,0 +1,418 @@
+         checked_cast_to_or_throw<uint8_t, Invalid_Argument>(scan_layer.arg_as_integer(0), "Invalid parameter");
+      const auto w =
+         checked_cast_to_or_throw<uint8_t, Invalid_Argument>(scan_layer.arg_as_integer(1), "Invalid parameter");
+      m_lms_lmots_params.push_back({LMS_Params::create_or_throw(hash, h), LMOTS_Params::create_or_throw(hash, w)});
+   }
+   m_max_sig_count = calc_max_sig_count();
+}
+
+HSS_Sig_Idx HSS_LMS_Params::calc_max_sig_count() const {
+   uint32_t total_height_counter = 0;
+   for(HSS_Level level(0); level < L(); level++) {
+      total_height_counter += params_at_level(level).lms_params().h();
+   }
+   if(total_height_counter >= sizeof(HSS_Sig_Idx) * 8) {
+      return HSS_Sig_Idx(std::numeric_limits<HSS_Sig_Idx::wrapped_type>::max());
+   }
+   return HSS_Sig_Idx(1) << total_height_counter;
+}
+
+HSS_LMS_PrivateKeyInternal::HSS_LMS_PrivateKeyInternal(const HSS_LMS_Params& hss_params, RandomNumberGenerator& rng) :
+      m_hss_params(hss_params), m_current_idx(0), m_sig_size(HSS_Signature::size(m_hss_params)) {
+   m_hss_seed = rng.random_vec<LMS_Seed>(m_hss_params.params_at_level(HSS_Level(0)).lms_params().m());
+   m_identifier = rng.random_vec<LMS_Identifier>(LMS_IDENTIFIER_LEN);
+}
+
+std::shared_ptr<HSS_LMS_PrivateKeyInternal> HSS_LMS_PrivateKeyInternal::from_bytes_or_throw(
+   std::span<const uint8_t> key_bytes) {
+   if(key_bytes.size() < sizeof(HSS_Level) + sizeof(HSS_Sig_Idx)) {
+      throw Decoding_Error("Too few private key bytes.");
+   }
+   BufferSlicer slicer(key_bytes);
+
+   const auto L = load_be<HSS_Level>(slicer.take<sizeof(HSS_Level)>());
+   if(L == 0U || L > HSS_MAX_LEVELS) {
+      throw Decoding_Error("Invalid number of HSS layers in private HSS-LMS key.");
+   }
+
+   const auto sig_idx = load_be<HSS_Sig_Idx>(slicer.take<sizeof(HSS_Sig_Idx)>());
+
+   std::vector<HSS_LMS_Params::LMS_LMOTS_Params_Pair> params;
+   for(size_t layer = 1; layer <= L; ++layer) {
+      if(slicer.remaining() < sizeof(LMS_Algorithm_Type) + sizeof(LMOTS_Algorithm_Type)) {
+         throw Decoding_Error("Out of bytes while parsing private HSS-LMS key.");
+      }
+      const auto lms_type = load_be<LMS_Algorithm_Type>(slicer.take<sizeof(LMS_Algorithm_Type)>());
+      const auto lmots_type = load_be<LMOTS_Algorithm_Type>(slicer.take<sizeof(LMOTS_Algorithm_Type)>());
+      params.push_back({LMS_Params::create_or_throw(lms_type), LMOTS_Params::create_or_throw(lmots_type)});
+   }
+   std::string hash_name = params.at(0).lms_params().hash_name();
+   if(!std::all_of(params.begin(), params.end(), [&hash_name](HSS_LMS_Params::LMS_LMOTS_Params_Pair& lms_lmots_params) {
+         bool valid_lmots_hash = lms_lmots_params.lmots_params().hash_name() == hash_name;
+         bool valid_lms_hash = lms_lmots_params.lms_params().hash_name() == hash_name;
+         return valid_lmots_hash && valid_lms_hash;
+      })) {
+      throw Decoding_Error("Inconsistent hash functions are not allowed.");
+   }
```

Comment:
Perhaps use `std::any_of` and reverse the predicate?

Reply:
fair